File: development/NET/samples/client/AECondition/VB/MyAeSubscription.vb
Imports System
Imports System.Windows.Forms
Imports Softing.OPCToolbox.Client
Imports Softing.OPCToolbox

Namespace AECondition
	Public Class MyAeSubscription
		Inherits AeSubscription

#Region "Constructor"
		'-----------------

		Public Sub New(ByVal parentSession As MyAeSession)
			MyBase.New(parentSession)
			AddHandler AeConditionsChanged, AddressOf HandleAeConditionsChanged

		End Sub

		'--
#End Region

#Region "Private Members"
		'---------------------


#Region "Private Attributes"
		'------------------------

		Private m_opcForm As OpcForm = Nothing

		'--
#End Region


		'--
#End Region

#Region "Public Methods"
		'---------------------

		Public Sub SetForm(ByVal form As OpcForm)
			m_opcForm = form
		End Sub

		Private Function GetState(ByVal state As EnumConditionState) As String
			Dim stateToString As String = String.Empty
			If (state And EnumConditionState.ACTIVE) = EnumConditionState.ACTIVE Then
				stateToString &= " ACT"
			End If
			If (state And EnumConditionState.ENABLED) = EnumConditionState.ENABLED Then
				stateToString &= " ENA"
			End If
			If (state And EnumConditionState.ACKED) = EnumConditionState.ACKED Then
				stateToString &= " ACK"
			End If
			If state = 0 Then
				stateToString &= " DIS"
			End If
			Return stateToString
		End Function

		'--
#End Region

#Region "Public Properties"
		'-----------------------


		'--
#End Region

#Region "Handles"
		'-----------------------

		Private Sub HandleAeConditionsChanged(ByVal anAeSubscription As AeSubscription, ByVal conditions As Client.AeCondition())

			If m_opcForm.conditionlistView.InvokeRequired Then
				Dim riflw As New RemoveItemsFromListView(AddressOf m_opcForm.RemoveItems)
				m_opcForm.conditionlistView.BeginInvoke(riflw, Nothing)
			Else
				m_opcForm.conditionlistView.Items.Clear()
			End If
			Dim state As String = [String].Empty
			Dim i As Integer = 0
			While i < conditions.Length
				Dim condition As Client.AeCondition = conditions(i)
				state = GetState(condition.State)

				If m_opcForm.conditionlistView.InvokeRequired Then
					Dim cclw As New ChangeConditionListView(AddressOf m_opcForm.ChangeItem)
					m_opcForm.conditionlistView.BeginInvoke(cclw, New Object() {condition, state})
				Else

					Dim item As New ListViewItem(state)
					item.Tag = condition
					m_opcForm.conditionlistView.Items.Add(item)
					item.SubItems.Add(condition.SourcePath)
					item.SubItems.Add(condition.ConditionName)
					item.SubItems.Add(condition.Severity.ToString())
					item.SubItems.Add(condition.Message)
					item.SubItems.Add(condition.OccurenceTime.ToString("hh:mm:ss.fff"))
					item.SubItems.Add(condition.ActorId)
					item.SubItems.Add(condition.SubConditionName)
					state = [String].Empty
				End If
				System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
			End While
			System.Diagnostics.Debug.WriteLine("The conditions list has changed ")
		End Sub

		'--
#End Region

		Public Delegate Sub ChangeConditionListView(ByVal condition As Client.AeCondition, ByVal state As String)
		Public Delegate Sub RemoveItemsFromListView()

	End Class

End Namespace

File: development/NET/samples/client/AECondition/CS/MyAeSubscription.cs
using System;
using Softing.OPCToolbox.Client;
using Softing.OPCToolbox;
using System.Windows.Forms;

namespace AECondition
{
	public class MyAeSubscription : AeSubscription
	{

		#region Constructor
		//-----------------

		public MyAeSubscription(MyAeSession parentSession) : base(parentSession)
		{
			AeConditionsChanged += new AeConditionsChangedEventHandler(HandleAeConditionsChanged);

		}

		//--
		#endregion

		#region Private Members
		//---------------------


		#region Private Attributes
		//------------------------

		private OpcForm m_opcForm = null;

		//--
		#endregion


		//--
		#endregion

		#region Public Methods
		//---------------------

		public void SetForm(OpcForm form)
		{
			m_opcForm = form;
		}

		private string GetState(EnumConditionState state)
		{
			string stateToString = string.Empty;
			if ((state & EnumConditionState.ACTIVE) == EnumConditionState.ACTIVE)
			{
				stateToString += " ACT";
			}
			if ((state & EnumConditionState.ENABLED) == EnumConditionState.ENABLED)
			{
				stateToString += " ENA";
			}
			if ((state & EnumConditionState.ACKED) == EnumConditionState.ACKED)
			{
				stateToString += " ACK";
			}
			if (state == 0)
			{
				stateToString += " DIS";
			}
			return stateToString;
		}

		//--
		#endregion

		#region Public Properties
		//-----------------------


		//--
		#endregion

		#region Handles
		//-----------------------

		private void HandleAeConditionsChanged(AeSubscription anAeSubscription, AeCondition[] conditions)
		{

			if (m_opcForm.conditionlistView.InvokeRequired)
			{
				RemoveItemsFromListView riflw = new RemoveItemsFromListView(m_opcForm.RemoveItems);
				m_opcForm.conditionlistView.BeginInvoke(riflw, null);
			}
			else
			{
				m_opcForm.conditionlistView.Items.Clear();
			}
			string state = String.Empty;
			for (int i = 0; i < conditions.Length; i++)
			{
				AeCondition condition = conditions[i];
				state = GetState(condition.State);

				if (m_opcForm.conditionlistView.InvokeRequired)
				{
					ChangeConditionListView cclw = new ChangeConditionListView(m_opcForm.ChangeItem);
					m_opcForm.conditionlistView.BeginInvoke(cclw, new object[] { condition, state });
				}
				else
				{

					ListViewItem item = new ListViewItem(state);
					item.Tag = condition;
					m_opcForm.conditionlistView.Items.Add(item);
					item.SubItems.Add(condition.SourcePath);
					item.SubItems.Add(condition.ConditionName);
					item.SubItems.Add(condition.Severity.ToString());
					item.SubItems.Add(condition.Message);
					item.SubItems.Add(condition.OccurenceTime.ToString("hh:mm:ss.fff"));
					item.SubItems.Add(condition.ActorId);
					item.SubItems.Add(condition.SubConditionName);
					state = String.Empty;
				}
			}//end for
			System.Diagnostics.Debug.WriteLine("The conditions list has changed ");
		}

		//--
		#endregion

		public delegate void ChangeConditionListView(AeCondition condition, string state);
		public delegate void RemoveItemsFromListView();

	}   //	end class MyAeItem

}   //	end namespace
